switchy/protocol.py

```
@@ -0,0 +1,267 @@
+            event.update(chunk)
+
+            iframe = s+2
+            clen = chunk.get('Content-Length')
+            if clen:
+                contents, segmented, diff, iframe = self.read_contents(
+                    data, iframe, clen)
+                if segmented:
+                    self._segmented = event, diff, contents
+                    break
+
+                if contents:
+                    event.update(self.parse_frame(contents))
+
+            events.append(event)
+            event = {}
+            s = data.find('\n\n', iframe)
+        else:
+            remaining = data[iframe:]
+            if remaining:  # segmented non-contents frame
+                self._segmented = event, 0, remaining
+
+        self.process_events(events, parsed)
+        return events  # for testing
+
+    def send(self, data):
+        """Write raw data to the transport.
+        """
+        msg = (data + '\n'*2).encode()
+        self.log.log(utils.TRACE, 'Data sent: {!r}'.format(msg))
+        self.transport.write(msg)
+
+    def sendrecv(self, data, resp_type='command/reply', fut=None):
+        """Send raw data to the transport and return a future representing
+        a response.
+        """
+        if not self.connected():
+            raise ConnectionError("Protocol is not connected")
+        fut = self.reg_fut(resp_type, fut=fut)
+        self.send(data)
+        return fut
+
+    def _handle_cmd_resp(self, future):
+        event = future.result()
+        resp = event.get('Body', event.get('Reply-Text', ''))
+        if not resp:
+            raise RuntimeError("Missing a response?")
+        lines = resp.splitlines()
+        if lines and '-ERR' in lines[-1]:
+            self.log.error("Event {} reported\n{}".format(event, resp))
+        return event
+
+    def bgapi(self, cmd, errcheck=True):
+        future = self.sendrecv('bgapi {}'.format(cmd))
+        if errcheck:
+            future.add_done_callback(self._handle_cmd_resp)
```

Why would `errcheck` be optional?

---

In case you want errors to fail silently.
This is mostly to be compat with `api`.
The original code didn't check for errors and so this was introduced later.

---

Ah, well, I'd suggest TODO'ing this and fixing this API when you break API with a major version number. Leave it to the caller to suppress errors.

---

Will do!